### ISSUE TITLE
FSDP 与 Megatron 后端的 grad_norm 存在显著差异

### ISSUE THREAD

**Issue by Eisenhower** [verl-project/verl#4710]
### 问题描述

在使用相同的模型如（Qwen3-1.7B）、相同的训练数据和超参数配置的情况下，从TensorBoard看，FSDP 和 Megatron在entropy、grad_norm上存在显著差异。从观测来看，grad_norm数值megatron普遍比fsdp2的大三倍左右。

### 环境信息

- **verl 版本**: v0.6.1、main latest
- **PyTorch 版本**: 2.8.0+cu128
- **Megatron-LM 版本**: 0.14.0rc7

### 复现配置

两个框架使用相同的超参数配置：

### 通用配置
model: Qwen3-1.7B
dp_size: 2
tp_size: 1 (Megatron)
pp_size: 1 (Megatron)
cp_size: 1 (Megatron)

### 训练超参数
ppo_mini_batch_size: 6
ppo_micro_batch_size_per_gpu: 1
gradient_accumulation: 6
use_dynamic_bsz: false
grad_clip: 1.0

### 问题截图

<img width="1309" height="592" alt="Image" src="https://github.com/user-attachments/assets/cffeb03f-de36-49af-b28a-8657d0051533" />

### 问题
1. 这是预期行为还是实现 bug？
2. 两个框架的 loss 计算流程是否存在已知差异？
3. 如何正确对比两个框架的训练效果？

### 期望行为
相同配置下，两个框架的 grad_norm 应该在相近数量级。

**Comment by MilesPoupart** [comment]
+1

<img width="1092" height="742" alt="Image" src="https://github.com/user-attachments/assets/514e7474-aaaa-49c1-8736-9db299f924c6" />

**Comment by Kirrito-k423** [comment]
中文：
感谢报告。基于目前公开信息，我倾向于先标记为需要更多信息，而不是直接判定为预期行为或实现 bug。

我查到几处相关证据：当前 `BaseEngine.train_batch()` 会把各 backend 的 `optimizer_step()` 返回值统一写入 `metrics["grad_norm"]`：https://github.com/verl-project/verl/blob/f663282327d784068263c7c3736a4884830eea44/verl/workers/engine/base.py#L124-L132 。FSDP 路径会先统计全局 `batch_num_tokens`/`dp_size`，然后通过 FSDP/FSDPModule 的 clip grad norm 路径返回 clip 前 norm：https://github.com/verl-project/verl/blob/f663282327d784068263c7c3736a4884830eea44/verl/workers/engine/fsdp/transformer_impl.py#L671-L682 和 https://github.com/verl-project/verl/blob/f663282327d784068263c7c3736a4884830eea44/verl/workers/engine/fsdp/transformer_impl.py#L730-L776 。Megatron 路径则返回 Megatron optimizer 的 `step()` 给出的 `grad_norm`，并且当前 engine 路径有 `batch_num_tokens`/`dp_size` 以及 `loss * num_micro_batch` 的归一化/PP schedule 补偿逻辑：https://github.com/verl-project/verl/blob/f663282327d784068263c7c3736a4884830eea44/verl/workers/engine/megatron/transformer_impl.py#L529-L548 、https://github.com/verl-project/verl/blob/f663282327d784068263c7c3736a4884830eea44/verl/workers/engine/megatron/transformer_impl.py#L675-L681 、https://github.com/verl-project/verl/blob/f663282327d784068263c7c3736a4884830eea44/verl/workers/engine/megatron/transformer_impl.py#L1116-L1123 。

同时，历史上确实有相近问题：#3655 讨论过 FSDP2 `grad_norm` 上报口径可能不是全 rank 聚合值，#5863/#5867 讨论过 legacy Megatron worker 的 loss normalization parity 问题。但 #5867 后续被关闭，理由是 legacy worker 已 deprecated，不再接收 PR。

为了判断这个 issue 是否仍是当前 main 的 bug，能否补充以下信息？
1. 两组实验的完整启动命令和完整 actor 配置，尤其是 `actor.strategy`/backend、`loss_agg_mode`、`use_remove_padding`、`use_dynamic_bsz`、`ppo_mini_batch_size`、`ppo_micro_batch_size_per_gpu`、`global_batch_size`、Megatron `use_mbridge`/是否 legacy worker。
2. 两组实验分别使用的 verl commit SHA，而不只是 `v0.6.1` 或 `main latest`。
3. TensorBoard 中对比的具体 metric key，是否都是 actor update 后的 `grad_norm`。
4. 一个最小可复现脚本或可公开的配置片段，最好能固定 seed，并只跑 2-3 个 actor update step。
5. 如果方便，请同时贴出同一步的 loss、entropy、KL、response token 数/`loss_mask.sum()`，这样可以区分是度量上报口径问题、loss normalization 问题，还是实际训练轨迹已经分叉。

English:
Thanks for the report. Based on the public evidence available so far, I would treat this as needing more information rather than immediately classifying it as either expected behavior or a confirmed implementation bug.

Relevant evidence: current `BaseEngine.train_batch()` records whatever each backend returns from `optimizer_step()` as `metrics["grad_norm"]`: https://github.com/verl-project/verl/blob/f663282327d784068263c7c3736a4884830eea44/verl/workers/engine/base.py#L124-L132 . The FSDP path computes global `batch_num_tokens`/`dp_size` and returns the pre-clipping norm through the FSDP/FSDPModule clip-grad path: https://github.com/verl-project/verl/blob/f663282327d784068263c7c3736a4884830eea44/verl/workers/engine/fsdp/transformer_impl.py#L671-L682 and https://github.com/verl-project/verl/blob/f663282327d784068263c7c3736a4884830eea44/verl/workers/engine/fsdp/transformer_impl.py#L730-L776 . The Megatron path returns the `grad_norm` from Megatron optimizer `step()`, and the current engine path includes `batch_num_tokens`/`dp_size` plus `loss * num_micro_batch` compensation for the Megatron PP schedule: https://github.com/verl-project/verl/blob/f663282327d784068263c7c3736a4884830eea44/verl/workers/engine/megatron/transformer_impl.py#L529-L548 , https://github.com/verl-project/verl/blob/f663282327d784068263c7c3736a4884830eea44/verl/workers/engine/megatron/transformer_impl.py#L675-L681 , and https://github.com/verl-project/verl/blob/f663282327d784068263c7c3736a4884830eea44/verl/workers/engine/megatron/transformer_impl.py#L1116-L1123 .

There are also related historical reports: #3655 discussed FSDP2 `grad_norm` reporting not being aggregated over all ranks, and #5863/#5867 discussed loss-normalization parity issues in the legacy Megatron worker. However, #5867 was closed because the legacy worker is deprecated and no longer accepts PRs.

To determine whether this is still a current-main bug, could you please provide:
1. The full launch commands and full actor configs for both runs, especially `actor.strategy`/backend, `loss_agg_mode`, `use_remove_padding`, `use_dynamic_bsz`, `ppo_mini_batch_size`, `ppo_micro_batch_size_per_gpu`, `global_batch_size`, Megatron `use_mbridge`, and whether the legacy worker is used.
2. The exact verl commit SHA for both runs, not only `v0.6.1` or `main latest`.
3. The exact TensorBoard metric key being compared, and whether both values are the actor-update `grad_norm`.
4. A minimal reproducible script or public config snippet, ideally with a fixed seed and only 2-3 actor update steps.
5. If possible, the same-step loss, entropy, KL, response-token count / `loss_mask.sum()`, so we can separate metric-reporting differences from loss-normalization differences or real training-trajectory divergence.

This response was generated by a scheduled Codex GPT5.5 issue processing task.
